platform/engineering/design-docs/2021-08-01-autonomous-deployments.md

```diff
@@ -0,0 +1,220 @@
+
+This work should be completed in 2-3 sprints with 2 people.
+
+### Alternatives
+
+An alternative that was discussed was isolating the applications in their own repositories before executing autonomous deploys. We felt that this would bloat the timeline and delay the feature from being released for too long. Isolating each application entails all of the groundwork needed to complete the autonomous deployment initiative, but also revamping deployments further, dependency isolation, foundational upgrades to the infrastructure, and large/disruptive changes to VFS teams' workflow as well as extensive training. While full isolation is a long term goal, it will be rolled out as part of an iterative plan; autonomous deployments being the first objective to complete.
+
+We also discussed monorepo isolation, but deemed it as unnecessary to complete the objective. Monorepo setup involves a technology infusion with something like Lerna, Yarn Workspaces, and/or NX. It was decided that pulling in new infrastructure at this stage would not get us any closer to the primary objective. This was also identified as an important upgrade to the system, but as it is non-critical, it is to be added to the roadmap as a follow-up initiative. Before this work can be completed, we need to make foundational upgrades to the platform code, architecture, and design system.
+
+One last option was to leave the system as-is, which the current deployment model. This was not pursued because it would not meet the objective. Another option was to keep everything we have in place right now, but require full builds and deploys for each merge into master. Ultimately we felt the deployment process would be too slow and prone to failure if we required full build and test suites for each deployment, which would dramatically increase strain on support.
+
+### Post MVP
+
+Initial dashboard view where teams sign in with Github auth, view their apps and manage their deployments including: scheduling, reverting, and manual deployment. This dashboard and infrastructure should be scoped and estimated separately as a post-MVP item. It will be the basis for a future platform dashboard.
+
+The goal of this post-MVP stage is to abstract the Github UI from the deployment actions so users are unaware of what is under the hood.
+
+After autonomous deployment work has been completed, work on application isolation will begin and another design document will be created.
+
+We would also like to be able to track which version of an app is currently released.
+
+### Questions
+- Should teams be able to merge into `master` without merging into `staging`/`dev`? Maybe only `hotfix/` and `revert/` branches can pass through?
+- What is the turnaround time to revert a commit?
+- Can we let teams deploy a full app or should be point to commits instead?
+- Should there be a maximum deployment frequency so a single app doesn't clog up the system?
+
```

Comment:
What happens when two apps try to deploy at the same time? For the MVP, I would certainly want them to get into a queue.

Reply:
They should be independent and just go out since we would be out of jenkins and just uploading files to S3